### PR TITLE
support querying without generating prepared statements

### PR DESCRIPTION
**Pull request by james-lawrence**
work related to #331, as a starting point for discussion around how to achieve making a query in pgx without passing through a prepared statement.

### CODE REVIEW

**Comment by james-lawrence** [comment]
this doesn't get us entirely to being able to connect to PG bouncer due to how the connection setup works. but I have a follow up PR that brings it the rest of the way.

**Comment by james-lawrence** [comment]
@jackc this is ready for a review.

fixes #331 probably fixes #344, and allows basic use with pgbouncer.

also lays a better foundation for supporting pg-like databases.

**Comment by jackc** [comment]
First of all, thanks for working on this.

There's a lot of changes here, and I'm not sure I get how they all fit together. It feels like there might really be several distinct PRs / changes here that are bundled together.

With that said, here are my thoughts so far.

1. Regarding moving pgx.TextFormatCode and pgx.BinaryFormatCode to pgtype package: this would be a breaking change for any code relying on it existing in the pgx package. So if we need those constants available in pgtype we would need to copy them and leave the originals alone. That could also reduce the size of the changeset as all the existing internal references to them could remain unchanged. I would probably make them unexported constants of pgtype so as not to increase the surface area of the package. In addition, it appears the reason for this was to add the preferred format code to DataType. While this might be an interesting refactor I didn't see anywhere that it was a prerequisite to any of the other features. So if I am correct that it isn't required by the other features I would prefer to consider this as a separate PR. (Another possibility -- [Preferred]FormatCode could be a method of DataType -- it would subsume the DetermineFormatCode reducing the interface of the package as well)

2. I haven't used PgBouncer, so I might be missing something, but wouldn't `NewPGBouncerConnInfo` render PGBouncer connections unable to use any types that do not have hard-coded OIDs (hstore for example). I would expect it still to be possible to query the server for a list of types even when being proxied through PGBouncer.

3. I'd prefer not to expose a complete `*pgx.QueryExOptions` in `DriverConfig`. Every field except `SimpleProtocol` is query specific. I think it would be confusing to expose config options that could break the connection if they were used. But at a more fundamental level, forcing the entire connection to default to the simple protocol strikes me as something that should be implemented in the core pgx package and enabled / configured from the stdlib package. I just did a quick proof-of-concept in branch prefer-simple-protocol (ce7014d372f12f3e4533fa8b84a1c2e6f395f82d).

4. I like the idea of an extension point for PG-like connections. I _think_ I like the `CustomConnInfo` callback function idea, but this would also be a lot easier to reason about in isolation.

**Comment by james-lawrence** [comment]
>There's a lot of changes here, and I'm not sure I get how they all fit together. It feels like there might really be several distinct PRs / changes here that are bundled together.

Totally agree! I was originally thinking about splitting into 3 distinct parts. I just decided to dump everything on you to see what you liked vs didn't like. I prefer smaller PRs myself.

tl;dr; I'll split up the PR into distinct parts with your comments taken into consideration.

> Regarding moving pgx.TextFormatCode and pgx.BinaryFormatCode ....

you are correct, I noticed it while working on this work, if we're okay with leaving unexported constants in the pgtype package I can do the rest without it.

> haven't used PgBouncer, so I might be missing something...

the distinction is connecting to PGBouncer vs connecting through PGBouncer, this isn't about connecting through pgbouncer, which as you stated you would still be able to query the list of types.
its about connecting directly to pgbouncer to query information about the process from its internal metrics.

>  I'd prefer not to expose a complete *pgx.QueryExOptions

fair enough, this particular change was there just to get a working example, I expected some push back there but didn't want to guess on the direction you'd have preferred.

> I like the idea of an extension point for PG-like connections....

I'll pull it out as a separate PR, its causing the errors in travis I have the fallback logic to cratedb incorrectly implemented apparently and need to revisit it.
